pkg/controller/direct/datacatalog/entry_controller.go

```diff
@@ -0,0 +1,365 @@
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/config"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/common"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/directbase"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/registry"
+)
+
+func init() {
+	registry.RegisterModel(krm.DataCatalogEntryGVK, NewEntryModel)
+}
+
+func NewEntryModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
+	return &entryModel{config: *config}, nil
+}
+
+var _ directbase.Model = &entryModel{}
+
+type entryModel struct {
+	config config.ControllerConfig
+}
+
+func (m *entryModel) client(ctx context.Context, projectID string) (*api.Client, error) {
+	var opts []option.ClientOption
+
+	// projectID is now passed as argument
+
+	config := m.config
+
+	// the service requires that a quota project be set
+	if !config.UserProjectOverride || config.BillingProject == "" {
+		config.UserProjectOverride = true
+		config.BillingProject = projectID // Use passed projectID
+	}
+
+	opts, err := config.RESTClientOptions()
+	if err != nil {
+		return nil, err
+	}
+
+	gcpClient, err := api.NewRESTClient(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("building datacatalog entry client: %w", err)
+	}
+
+	return gcpClient, nil
+}
+
+func (m *entryModel) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) { // Reverted back to client.Reader
+	obj := &krm.DataCatalogEntry{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	id, err := krm.NewEntryIdentity(ctx, reader, obj)
+	if err != nil {
+		return nil, err
+	}
+
+	gcpClient, err := m.client(ctx, id.Parent().ProjectID)
+	if err != nil {
+		return nil, err
+	}
+
+	return &entryAdapter{
+		gcpClient: gcpClient,
+		id:        id,
+		desired:   obj,
+		reader:    reader,
+	}, nil
+}
+
+func (m *entryModel) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, nil
+}
+
+type entryAdapter struct {
+	gcpClient *api.Client
+	id        *krm.EntryIdentity
+	desired   *krm.DataCatalogEntry
+	actual    *pb.Entry
+	reader    client.Reader // Reverted back to client.Reader
+}
+
+var _ directbase.Adapter = &entryAdapter{}
+
+func (a *entryAdapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("getting datacatalog entry", "name", a.id)
+
+	req := &pb.GetEntryRequest{Name: a.id.String()}
+	actual, err := a.gcpClient.GetEntry(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		// PermissionDenied is also observed for non-existent entries
+		if direct.IsPermissionDenied(err) {
```

Comment:
Just to gather more context: why does API returns either NotFound or PermissionDenied for non-existent entries? How can we tell whether this is a true permission issue?

Reply:
From a security point of view i can see this happening to things like username or sensitive keys. To prevent scrappers from identifying which keys are available at the server (404 not available, 403 available but permission denied). Not sure why they are doing for this resource.

Reply:
In fact it does not return 404 at all. The code above this 403 check is mostly defensive coding.

Reply:
Understood. I guess we'll just try to create a new resource whenever the error is PermissionDenied.